runtime/src/iree/base/loop_emscripten.js

```diff
@@ -44,14 +47,112 @@ const IreeLibraryLoopEmscripten = {
       abort() {
         clearTimeout(this.timeoutId);
         Module['dynCall'](
-            'iiii', this.callback, this.user_data, this.loop,
-            IREE_STATUS_ABORTED);
+            'iiii', this.callback,
+            [this.user_data, this.loop, IREE_STATUS_ABORTED]);
+      }
+    }
+
+    // IREE_LOOP_COMMAND_WAIT_UNTIL
+    class LoopCommandWaitUntil extends LoopCommand {
+      constructor(scope, operationId, callback, user_data, timeout_ms, loop) {
+        super();
+
+        this.callback = callback;
+        this.user_data = user_data;
+        this.loop = loop;
+        this.abortFn = undefined;
+
+        const abortPromise = new Promise((_, reject) => {
+          this.abortFn = reject;
+        });
+        const timeoutPromise = new Promise((resolve, _) => {
+          setTimeout(() => {
+            resolve();
+          }, timeout_ms);
+        });
+
+        Promise.race([abortPromise, timeoutPromise])
+            .then(() => {
+              Module['dynCall'](
+                  'iiii', this.callback,
+                  [this.user_data, this.loop, IREE_STATUS_OK]);
+              // TODO(scotttodd): handle the returned status (sticky failure
+              //     state?) at least free the status so it doesn't leak
+              delete scope.pendingOperations[operationId];
+            })
+            .catch(() => {
+              Module['dynCall'](
+                  'iiii', this.callback,
+                  [this.user_data, this.loop, IREE_STATUS_ABORTED]);
+            })
+            .finally(() => {
+              delete scope.pendingOperations[operationId];
+            });
+      }
+
+      abort() {
+        this.abortFn();
+      }
+    }
+
+    // IREE_LOOP_COMMAND_WAIT_ONE
+    // IREE_LOOP_COMMAND_WAIT_ANY
+    // IREE_LOOP_COMMAND_WAIT_ALL
+    class LoopCommandWaitPromise extends LoopCommand {
+      constructor(
+          scope, operationId, callback, user_data, timeout_ms, wait_promise,
+          loop) {
+        super();
+
+        this.callback = callback;
+        this.user_data = user_data;
+        this.loop = loop;
+        this.abortFn = undefined;
+
+        // We're given a primary Promise to wait on, but we'll race it against
+        // two other Promises:
+        //   * timeoutPromise: rejected after a timeout
+        //   * abortPromise : rejected if abort() is called
+        // This way, whichever Promise settles first will decide the eventual
+        // state and either issue the callback with IREE_STATUS_OK or
+        // IREE_STATUS_ABORTED.
+        const abortPromise = new Promise((_, reject) => {
+          this.abortFn = reject;
+        });
+        const timeoutPromise = new Promise((_, reject) => {
```

Comment:
since most waits are of the INFINITE variety could can conditionally set this promise - no use making a bunch of browser calls to setTimeout(INFINITE) and race the promise that'll never resolve.